GillianCore/logging/logging.ml

```diff
@@ -1,129 +1,250 @@
-module Mode = Mode
-module Report = Report
+module Reporter = Reporter
+module TargetLang = TargetLang
 
-let wrap_up = Reporter.wrap_up
+module Dummy : TargetLang.S = struct
+  type t
 
-let log lvl msgf =
-  if Mode.should_log lvl then
-    let report = ReportBuilder.info "" (Debug msgf) () in
-    Reporter.log report
+  let reporters =
+    [ new Reporter.file_reporter (); new Reporter.database_reporter () ]
+end
 
-let normal msgf = log Normal msgf
+module type S = sig
```

Comment:
One solution you could adopt to reduce duplication is creating a `reporter_S.mli` which just contains the signature.
In the dune file you will need to add `(modules_without_implementation reporter_S)`
Then you should be able to say things like
```ocaml
module Make(P: TargetLang.S) : Reporter_S
module Default : Reporter_S
```

I find `Reporter_S` a little less elegant than `Reporter.S`, but code duplication is difficult to maintain